add column name normalization

alecloudenback opened the pull request:
In reference to #22 

Primary code taken from MIT licensed CSV.jl[1], originally written by @quinnj I think.

I am having a hard time getting test to run - I think I am not utilizing Snoop correctly. I get a lot of errors similar to this when trying to run the tests as is:

```julia
 MethodError: no method matching TableReader.ParserParameters(::Char, ::Char, ::Bool, ::Bool, ::Int64, ::Bool, ::Nothing, ::Bool, ::Bool, ::Int64)
```
I tried running the `/tools/snoop.jl` and copying the output to the end of the `TableReader.jl` file, but that didn't seem to work. Reading the documentation of SnoopCompile.jl, it seems like there should be a `/tmp/precompile` folder created that I should also bring into `/precompile/` but I don't see that folder get generated (I do get a `compiles.csv`). (I did not include the changes I attempted related to `Snoop.jl` in the initial commit).

Also, this is my first real attempt at an issue/pull request on GitHub, so please let me know if I'm going about it wrong.

1: https://github.com/JuliaData/CSV.jl/blob/da5bcbb53a3f4b02eb1d71b0849a5f88e1d12fe0/src/filedetection.jl#L8

alecloudenback committed:
add column name normalization

bicycle1885 commented:
Great! I think this feature is nice. Please go ahead. Let me add a few comments:

- Description of the normalizing behavior is required in the docs.
- If you take the code from another repository, you need to mention it and clarify the range of it.
- Precompiled files will be generated at the current directory where you are working; please check `precompile/precompile_TableReader.jl` file.

quinnj commented:
I think you need the `RESERVED` list of words, right? Here: https://github.com/JuliaData/CSV.jl/blob/master/src/filedetection.jl#L1. Also, don't worry about attributing here; I'm pretty sure I took a start from some [DataFrames](https://github.com/JuliaData/DataFrames.jl/blob/a8904f678947e8d589a7bbe616174f1db022b3e0/src/deprecated.jl#L770) and enhanced it for CSV.jl. Really, someone should just put this in Base since it's generally useful.

alecloudenback committed:
add reserved words for use in normalization

alecloudenback committed:
round out normalize names feature and more test cases

alecloudenback committed:
update Snoop.jl precompile

alecloudenback committed:
add a few comments and remove @show that was not needed

alecloudenback commented:
I think this should be good to go - I added some test cases, documentation in `TableReader.jl`, and updated the `precompile` block with the newly generated Snoop output. Please let me know if you have thoughts/feedback.

bicycle1885 commented on src/TableReader.jl:
```diff
@@ -232,13 +238,14 @@ for (fname, delim) in [(:readdlm, nothing), (:readcsv, ','), (:readtsv, '\t')]
     push!(kwargs, Expr(:kw, :(skip::Integer), 0))  # skip::Integer = 0
     push!(kwargs, Expr(:kw, :(skipblank::Bool), true))  # skipblank::Bool = true
     push!(kwargs, Expr(:kw, :(colnames), nothing))  # colnames = nothing
+    push!(kwargs, Expr(:kw, :(normalizenames), false))  # skipblank::Bool = false
```
Can you update the comment like `# normalizenames::Bool = false`?

alecloudenback replied:
this is fixed in the latest commit

bicycle1885 commented on src/TableReader.jl:
```diff
@@ -661,34 +673,34 @@ function _precompile_()
-    precompile(Tuple{getfield(TableReader, Symbol("##readcsv#12")), Char, Char, Bool, Int64, Bool, Nothing, Int64, typeof(identity), String})
-    precompile(Tuple{getfield(TableReader, Symbol("##readtsv#21")), Char, Char, Bool, Int64, Bool, Nothing, Int64, typeof(identity), String})
 end
-_precompile_()
```
Please do not remove this line. This precompiles methods.

alecloudenback replied:
fixed in latest commit

bicycle1885 commented on test/runtests.jl:
```diff
@@ -830,3 +830,25 @@ end
+        1,2,3
+        """)
+        df = readcsv(buffer,normalizenames=false)
+        @test names(df) == [:col1, :col2, :col3]
```
I think these column names are untouched even if `normalizenames = true`. So, this tests nothing. I think you should test the change of the behavior caused by `normalizenames` using the same column names.

alecloudenback replied:
the test case is enhanced in latest commit

alecloudenback committed:
add back precompile, better "don't normalize" test

bicycle1885 commented on src/parser.jl:
```diff
@@ -1,3 +1,5 @@
+using Unicode
```
I think you need to add this module to Project.toml by running `add Unicode`.

alecloudenback replied:
Unicode was already a part of the project, so no change to make I think

alecloudenback committed:
fix normalizenames comment

alecloudenback committed:
Update Readme.md

bicycle1885 commented:
Great work! This looks good. I will merge this and close #22.

alecloudenback commented:
Thanks for the help with the PR, @bicycle1885
